source/common/http/filter/gzip_filter.cc

```diff
@@ -0,0 +1,116 @@
+#include "common/http/filter/gzip_filter.h"
+
+#include <regex>
+
+namespace Envoy {
+namespace Http {
+
+GzipFilter::GzipFilter(GzipFilterConfigSharedPtr config)
+    : skip_compression_{true}, compressed_data_(), compressor_(), config_(config) {}
+
+void GzipFilter::onDestroy() {}
+
+FilterHeadersStatus GzipFilter::decodeHeaders(HeaderMap& headers, bool) {
+  // TODO(gsagula): The current implementation checks for the presence of 'gzip' and if the same
+  // is followed by Qvalue. Since gzip is the only available encoding right now, order/priority of
+  // preferred server encodings is disregarded (RFC2616-14.3).
+  skip_compression_ = !isAcceptEncodingGzip(headers);
+  return FilterHeadersStatus::Continue;
+}
+
+FilterHeadersStatus GzipFilter::encodeHeaders(HeaderMap& headers, bool end_stream) {
+  if (end_stream || skip_compression_) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  // TODO(gsagula): should also inspect transfer-encoding header.
+  if (headers.ContentEncoding() == nullptr && isContentTypeAllowed(headers) &&
+      isMinimumContentLength(headers) && isCacheControlAllowed(headers) && isEtagAllowed(headers) &&
+      isLastModifiedAllowed(headers)) {
+    headers.removeContentLength();
+    headers.insertContentEncoding().value(Http::Headers::get().ContentEncodingValues.Gzip);
+    compressor_.init(config_->getCompressionLevel(), config_->getCompressionStrategy(), 31,
+                     config_->getMemoryLevel());
+  } else {
+    skip_compression_ = true;
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+} // namespace Http
+
+FilterDataStatus GzipFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (skip_compression_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  const uint64_t n_data{data.length()};
+
+  if (n_data > 0) {
+    compressor_.compress(data, compressed_data_);
+  }
+
+  if (end_stream) {
+    compressor_.flush(compressed_data_);
+  }
+
+  if (compressed_data_.length() > 0) {
+    data.drain(n_data);
+    data.move(compressed_data_);
+    return Http::FilterDataStatus::Continue;
+  }
+
+  return Http::FilterDataStatus::StopIterationNoBuffer;
+}
+
+bool GzipFilter::isAcceptEncodingGzip(const HeaderMap& headers) const {
+  if (headers.AcceptEncoding() != nullptr) {
+    return std::regex_search(
+        headers.AcceptEncoding()->value().c_str(),
+        std::regex{"(?!.*gzip;\\s*q=0(,|$))(?=(.*gzip)|(^\\*$))", std::regex::optimize});
```

Comment:
My understanding is that `std::regex::optimize` flag makes matching faster at the expense of slower construction. Since this uses a temporary it'll still be doing construction for each call so this trade off doesn't seem to make sense. Thoughts?

Reply:
(I think ideally you could have this `std::regex` live the same lifespan as Envoy itself, but watch out for the "static initialization order fiasco")